### PR TITLE
#20080 first draft of the merge bulk streaming fire action

### PR DESCRIPTION
**Pull request by jdotcms**
First draft for the parallel bulk of contentlet merge action of firing a default action on streaming json response

### CODE REVIEW

**Comment by erickgonzalez** [comment]
Please improve/update the description.

We have a couple of tests failing:

https://raw.githack.com/dotCMS/test-results/issue-20080-workflow-patch-merge/projects/core/current/curl/postgres/reports/html/Workflow_Resource_Tests.html#pills-failed

https://raw.githack.com/dotCMS/test-results/issue-20080-workflow-patch-merge/projects/core/current/integration/postgres/reports/html/classes/com.dotcms.content.elasticsearch.business.ESContentletAPIImplTest.html#testCheckInWithLegacyRelationshipsAndReadOnlyClusterShouldThrowAnException

File: dotCMS/src/main/java/com/dotcms/concurrent/DotConcurrentFactory.java [hunk 1]
```diff
@@ -578,6 +656,91 @@ public void run() {
...
+        @Override
+        public <T> Future<T> submit(final Callable<T> callable) {
+
+            return this.executorService.submit(callable);
+        }
+
+        @Override
+        public <T> Future<T> submit(Callable<T> callable, long delay, TimeUnit unit) {
+
+            throw new UnsupportedOperationException("Submit Delay not supported on single submitter");
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/concurrent/DotConcurrentFactory.java:690]: Is it possible to add the name of the submitter in the error msg? For troubleshooting/support purposes.

File: dotCMS/src/main/java/com/dotcms/concurrent/DotConcurrentFactory.java [hunk 2]
```diff
@@ -578,6 +656,91 @@ public void run() {
...
+        public boolean isAborting() {
+            return this.executorService.isTerminated() ||  this.executorService.isShutdown();
+        }
+
+        @Override
+        public void waitForAll(long timeout, TimeUnit unit) throws ExecutionException {
+            try {
+                executorService.awaitTermination(timeout, unit);
+            } catch (InterruptedException e) {
+                throw new DotRuntimeException(e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/concurrent/DotConcurrentFactory.java:729]: Is it possible to add the name of the submitter in the error msg? For troubleshooting/support purposes.

File: dotCMS/src/main/java/com/dotcms/concurrent/DotConcurrentFactory.java [hunk 3]
```diff
@@ -578,6 +656,91 @@ public void run() {
...
+            }
+        }
+
+        @Override
+        public long getTaskCount() {
+            throw new UnsupportedOperationException("Submit Delay not supported on single submitter");
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/concurrent/DotConcurrentFactory.java:735]: Is it possible to add the name of the submitter in the error msg? For troubleshooting/support purposes.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 4]
```diff
@@ -4457,7 +4457,8 @@ private Contentlet internalCheckin(Contentlet contentlet,
         }
 
         if (!isCheckInSafe(contentRelationships)){
-            this.esReadOnlyMonitor.start();
+            Try.of(()->DotConcurrentFactory.getInstance().getSingleSubmitter().submit(()-> this.elasticReadOnlyCommand.executeCheck()))
+                    .onFailure(e -> Logger.error(ESContentletAPIImpl.class, e.getMessage()));
             throw new DotContentletStateException(
                     "Content cannot be saved at this moment. Reason: Elastic Search cluster is in read only mode.");
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:4463]: As an additional change, can you please add the content ID to the error message?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ElasticReadOnlyCommandImpl.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+            final SystemMessageBuilder messageBuilder = new SystemMessageBuilder()
+                    .setMessage(message)
+                    .setSeverity(MessageSeverity.ERROR)
+                    .setType(MessageType.SIMPLE_MESSAGE)
+                    .setLife(TimeUnit.SECONDS.toMillis(5));
+            Logger.error(this.getClass(), message);
+            systemMessageEventUtil.pushMessage(messageBuilder.create(), usersId);
+        } catch (final  LanguageException | DotDataException | DotSecurityException e) {
+            Logger.warn(this, () -> e.getMessage());
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ElasticReadOnlyCommandImpl.java:183]: Please add the message key to the error msg

File: dotCMS/src/main/java/com/dotmarketing/factories/PublishFactory.java [hunk 6]
```diff
@@ -589,7 +588,7 @@ private static void sendMessage(final User user, final String messageKey, String
 
 			Logger.warn(PublishFactory.class, message);
 		} catch (final  LanguageException  e) {
-			Logger.warn(ESReadOnlyMonitor.class, () -> e.getMessage());
+			Logger.warn(PublishFactory.class, () -> e.getMessage());
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/factories/PublishFactory.java:591]: Please add the message key to the error msg

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java [hunk 7]
```diff
@@ -128,36 +128,6 @@ public void markAsFailed() throws DotDataException {
...
-
-    /**
-     * Method to Test: {@link ReindexQueueAPIImpl#markAsFailed(ReindexEntry, String)}
-     * When: Try to mark a reindex as fail and a {@link DotDataException} in thrown
-     * Should: Should not call the {@link ESReadOnlyMonitor#start(String)} )} )} method
-     *
-     * @throws DotDataException
-     */
-    @Test
-    public void markAsFailedWhenAExceptionIsThrown() throws DotDataException {
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java:142]: what happens now with the monitor if an exception is thrown?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java [hunk 8]
```diff
@@ -567,23 +569,36 @@ private void indexContentListDefer(final List<Contentlet> contentToIndex) {
...
+                        Logger.error(this,
+                                "Erro" +
+                                        "r reindexing (" + response.getItems().length + ") content(s) "
+                                        + response.buildFailureMessage());
+                    }
                 }
             }
-
+        } catch(final IllegalStateException e) {
+            ContentletFactory.rebuildRestHighLevelClientIfNeeded(e);
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java:593]: This code can be removed as the catch Exception considers this case

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ElasticReadOnlyCommandImpl.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+        if (timeToWaitAfterWriteModeSet == TIME_TO_WAIT_AFTER_WRITE_MODE_SET_NOTY_VALUE) {
+            loadTimeToWaitAfterWriteModeSet();
+        }
+
+        final boolean clusterInReadOnlyMode              = Try.of(()-> ElasticsearchUtil.isClusterInReadOnlyMode())
+                .onFailure(e->Logger.warn(ElasticReadOnlyCommand.class,  "unable to access ES Cluster Metadata: " + e.getMessage()))
+                .getOrElse(true);
+        final boolean eitherLiveOrWorkingIndicesReadOnly = Try.of(()-> ElasticsearchUtil.isEitherLiveOrWorkingIndicesReadOnly())
+                .onFailure(e->Logger.warn(ElasticReadOnlyCommand.class,  "unable to access ES Index Metadata: " + e.getMessage()))
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ElasticReadOnlyCommandImpl.java:94]: is this log message correct? ES cluster metadata?

Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ElasticReadOnlyCommandImpl.java:94]: is this log message correct? ES cluster metadata?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ElasticReadOnlyCommandImpl.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+    @Override
+    public void executeCheck() {
+
+        if (timeToWaitAfterWriteModeSet == TIME_TO_WAIT_AFTER_WRITE_MODE_SET_NOTY_VALUE) {
+            loadTimeToWaitAfterWriteModeSet();
+        }
+
+        final boolean clusterInReadOnlyMode              = Try.of(()-> ElasticsearchUtil.isClusterInReadOnlyMode())
+                .onFailure(e->Logger.warn(ElasticReadOnlyCommand.class,  "unable to access ES Cluster Metadata: " + e.getMessage()))
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ElasticReadOnlyCommandImpl.java:91]: is this log message correct? ES cluster metadata?

File: dotCMS/src/test/java/com/dotcms/concurrent/DotConcurrentFactoryTest.java [hunk 11]
```diff
@@ -2,16 +2,90 @@
...
+        final DotConcurrentFactory dotConcurrentFactory =
+                DotConcurrentFactory.getInstance();
+        final DotSubmitter submitter =
+                dotConcurrentFactory.getSingleSubmitter();
+        submitter.submit(()-> esReadOnlyMonitor.executeCheck());
+
+        verify(esReadOnlyMonitor).executeCheck();
+    }
+
+    @Test
```
Comment (nollymar) [inline dotCMS/src/test/java/com/dotcms/concurrent/DotConcurrentFactoryTest.java:42]: Please add javadoc

File: dotCMS/src/curl-test/Workflow Resource Tests.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
 	"info": {
-		"_postman_id": "cd87df22-9b12-4ee0-b05f-290155476363",
+		"_postman_id": "fd766073-2359-456c-99ae-2563435c67e4",
```
Comment (nollymar) [inline dotCMS/src/curl-test/Workflow Resource Tests.json:3]: when you modified this file, did you make sure you were working with the latest version? it seems that some tests are being removed with your commit

File: dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java (hunk 7 above)
Comment (jdotcms) [reply dotCMS/src/integration-test/java/com/dotmarketing/common/reindex/ReindexQueueAPITest.java:142]: the monitor was removed from the Reindex thread in favor of a cron job

**Review by freddyucv — APPROVED** [review]

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESReadOnlyMonitorTest.java [hunk 13]
```diff
@@ -37,7 +37,7 @@
 public class ESReadOnlyMonitorTest {
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESReadOnlyMonitorTest.java:37]: This class should be renamed as the ESReadOnlyMonitor was removed and now the ElasticReadOnlyCommand is the class being tested

File: dotCMS/src/curl-test/Workflow Resource Tests.json (hunk 12 above)
Comment (jdotcms) [reply dotCMS/src/curl-test/Workflow Resource Tests.json:3]: some curl I have done, didn't make sense at the end

**Review by nollymar — APPROVED** [review]

**Pull request merged**